### PR TITLE
Implement contracted score panel and integrate into results screen

### PR DESCRIPTION
**Pull request by smoogipoo**
The changes look big at first, but there's one refactoring that touched a significant number of files. Rest is pure implementation.

Still using the default scroll container's scroll bar for this initial revision.

### CODE REVIEW

**Review by bdach — COMMENTED** [review]
This looks really *really* cool in game! 👍 

I have some comments but they're all nits. The rest looks good to me.

**Comment by bdach** [comment]
Test failure seems to be best addressed by providing a no-op override of `LoadComplete()` in [`FadeAccessibleResults`](https://github.com/ppy/osu/blob/9ca4fdf95bf27b10d552f67115db6f1cdf690ce4/osu.Game.Tests/Visual/Background/TestSceneUserDimBackgrounds.cs#L329-L339) to avoid it firing off any API requests whatsoever.

As a general point, are you planning to allow drag scrolling over the horizontal axis going forward? Right now dragging seems sort of wrong as it goes vertically but not horizontally, so you have to exclusively use the scrollbar or click on multiple items to see stuff that's off screen, which feels off.

**Comment by smoogipoo** [comment]
> As a general point, are you planning to allow drag scrolling over the horizontal axis going forward? 

Maybe, but I don't think so. Scrolling downward will reveal additional stats, similar to stable but with more data. Allowing horizontal drag complicates that, especially on touch devices.

> Right now dragging seems sort of wrong

This is probably more a result of the scrollbar existing in the current raw state. The full scrollbar design is going to look something like:
image
And will be locked to the bottom of the screen.

File: osu.Game.Rulesets.Osu/Scoring/OsuScoreProcessor.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 namespace osu.Game.Rulesets.Osu.Scoring
 {
-    internal class OsuScoreProcessor : ScoreProcessor
+    public class OsuScoreProcessor : ScoreProcessor
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Scoring/OsuScoreProcessor.cs:11]: This was done for dynamic compilation. We should completely avoid `internal` where it's not 100% necessary.

File: osu.Game/Screens/Ranking/ScorePanelList.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+            // Add padding to both sides such that the centre of an expanded panel on either side is in the middle of the screen.
+            flow.Padding = new MarginPadding { Horizontal = DrawWidth / 2f - ScorePanel.EXPANDED_WIDTH / 2f - expanded_panel_spacing };
+        }
+
+        private class Flow : FillFlowContainer<ScorePanel>
+        {
+            public override IEnumerable<Drawable> FlowingChildren => applySorting(AliveInternalChildren);
+
+            public int GetPanelIndex(ScoreInfo score) => applySorting(Children).OfType<ScorePanel>().TakeWhile(s => s.Score != score).Count();
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Ranking/ScorePanelList.cs:121]: Should be impossible since we're never removing panels, at least not for now. Maybe fine to just keep a note of for if that ever happens.

**Comment by smoogipoo** [comment]
> Test failure seems to be best addressed by providing a no-op override of LoadComplete() in FadeAccessibleResults to avoid it firing off any API requests whatsoever.

I've resolved a bit differently by moving the abstraction I had planned back into this branch and putting the request inside `SoloResultsScreen`.

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
Not 100% with some pieces but we can fix those up in future refactors. Main things I'd look at going forward:

- Instant flow size changes feels quite jerky
- Really need that horizontal scroll to make it feel good
- The scroll logic is a bit hard to follow, but commented and local so I'm not going to be too picky

**Comment by smoogipoo** [comment]
> Instant flow size changes feels quite jerky

Can you describe what part of it? Half the changes I made and the reason for the complexity in the scroll logic was to completely remove the jerkiness. Unless you mean something else?

**Comment by peppy** [comment]
When clicking another panel, the flow changes size immediately, causing a sudden jump in the underlying panels:

image

(played at 1% speed to increase visibility)

An easy fix may be to transform the flow to the new state.

**Comment by smoogipoo** [comment]
Interesting, I didn't notice that in my testing. I made the size transform instantly to fix other scroll problems but that's probably not required anymore.

Will keep it in mind.
